RVV1.0 Supported tests for RISC-V

alitariq4589 opened the pull request:
This PR adds tests supported for RISC-V

## Tests which are added for execution with RISC-V

`debian-cpu-cmake-rv64-native`
`debian-trixie-cmake-sanitizer-riscv64-native` * 3
`debian-trixie-llguidance-riscv64-native`
`debian-trixie-cmake-rpc-riscv64-native`
`ggml-ci-riscv64-native-cpu-low-perf`


## Dependencies which are not yet supported for RISC-V:

`rocblas-dev`
`hipblas-dev`
`vulkan-sdk`
`mthreads/musa:rc4.3.0-devel-ubuntu22.04-amd64`
`intel-oneapi-compiler-dpcpp-cpp`
`intel-oneapi-mkl-devel`
`cuda`
`macOS*`
`windows*`
`torch`


## Tests which are not added for RISC-V due to above unmet dependencies:

`macOS-latest-cmake-arm64`
`macOS-latest-cmake-x64`
`macOS-latest-cmake-arm64-webgpu`
`ubuntu-24-cmake-vulkan`
`ubuntu-22-cmake-webgpu`
`ubuntu-22-cmake-hip`
`ubuntu-22-cmake-musa`
`ubuntu-22-cmake-sycl`
`ubuntu-22-cmake-sycl-fp16`
`macOS-latest-cmake-ios`
`macOS-latest-cmake-tvos`
`macOS-latest-cmake-visionos`
`macOS-latest-swift`
`windows-msys2`
`windows-latest-cmake`
`ubuntu-latest-cmake-cuda`
`windows-2022-cmake-cuda`
`windows-latest-cmake-sycl`
`windows-latest-cmake-hip`
`ios-xcode-build`
`android-build`
`openEuler-latest-cmake-cann`
`ggml-ci-x64-nvidia-cuda`
`ggml-ci-x64-nvidia-vulkan-cm`
`ggml-ci-x64-nvidia-vulkan-cm2`
`ggml-ci-x64-cpu-amx`
`ggml-ci-mac-metal`
`ggml-ci-mac-vulkan`
`ggml-ci-arm64-cpu-high-perf-sve`
`ggml-ci-riscv64-native-cpu-high-perf`

## Additional Notes

### Note 1

Due to a warning (treated as error) related to RISC-V simd mappings

```
In file included from ../../../ggml/src/ggml-cpu/arch/riscv/quants.c:6:
../../../ggml/src/ggml-cpu/simd-mappings.h: In function 'riscv_compute_fp16_to_fp32':
../../../ggml/src/ggml-cpu/simd-mappings.h:101:9: error: ISO C does not support the '_Float16' type before C23 [-Werror=pedantic]
  101 |         _Float16 hf;
      |         ^~~~~~~~
../../../ggml/src/ggml-cpu/simd-mappings.h: In function 'riscv_compute_fp32_to_fp16':
../../../ggml/src/ggml-cpu/simd-mappings.h:108:9: error: ISO C does not support the '_Float16' type before C23 [-Werror=pedantic]
  108 |         _Float16 hf = (_Float16)f;
      |         ^~~~~~~~
../../../ggml/src/ggml-cpu/simd-mappings.h:108:24: error: ISO C does not support the '_Float16' type before C23 [-Werror=pedantic]
  108 |         _Float16 hf = (_Float16)f;
      |                        ^~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [ggml/src/CMakeFiles/ggml-cpu.dir/build.make:261: ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/arch/riscv/quants.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from ../../../ggml/src/ggml-cpu/quants.c:5:
../../../ggml/src/ggml-cpu/simd-mappings.h: In function 'riscv_compute_fp16_to_fp32':
../../../ggml/src/ggml-cpu/simd-mappings.h:101:9: error: ISO C does not support the '_Float16' type before C23 [-Werror=pedantic]
  101 |         _Float16 hf;
      |         ^~~~~~~~
../../../ggml/src/ggml-cpu/simd-mappings.h: In function 'riscv_compute_fp32_to_fp16':
../../../ggml/src/ggml-cpu/simd-mappings.h:108:9: error: ISO C does not support the '_Float16' type before C23 [-Werror=pedantic]
  108 |         _Float16 hf = (_Float16)f;
      |         ^~~~~~~~
../../../ggml/src/ggml-cpu/simd-mappings.h:108:24: error: ISO C does not support the '_Float16' type before C23 [-Werror=pedantic]
  108 |         _Float16 hf = (_Float16)f;
      |                        ^~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [ggml/src/CMakeFiles/ggml-cpu.dir/build.make:135: ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/quants.c.o] Error 1
In file included from ../../../ggml/src/ggml-cpu/common.h:7,
                 from ../../../ggml/src/ggml-cpu/unary-ops.h:3,
                 from ../../../ggml/src/ggml-cpu/ggml-cpu.c:12:
../../../ggml/src/ggml-cpu/simd-mappings.h: In function 'riscv_compute_fp16_to_fp32':
../../../ggml/src/ggml-cpu/simd-mappings.h:101:9: error: ISO C does not support the '_Float16' type before C23 [-Werror=pedantic]
  101 |         _Float16 hf;
      |         ^~~~~~~~
../../../ggml/src/ggml-cpu/simd-mappings.h: In function 'riscv_compute_fp32_to_fp16':
../../../ggml/src/ggml-cpu/simd-mappings.h:108:9: error: ISO C does not support the '_Float16' type before C23 [-Werror=pedantic]
  108 |         _Float16 hf = (_Float16)f;
      |         ^~~~~~~~
../../../ggml/src/ggml-cpu/simd-mappings.h:108:24: error: ISO C does not support the '_Float16' type before C23 [-Werror=pedantic]
  108 |         _Float16 hf = (_Float16)f;
      |                        ^~~~~~~~
../../../ggml/src/ggml-cpu/ggml-cpu.c: In function 'ggml_cpu_fp32_to_fp16':
../../../ggml/src/ggml-cpu/ggml-cpu.c:3239:32: error: ISO C does not support the '_Float16' type before C23 [-Werror=pedantic]
 3239 |         __riscv_vse16_v_f16m1((_Float16 *)&y[i], vy, vl);
      |                                ^~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [ggml/src/CMakeFiles/ggml-cpu.dir/build.make:79: ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/ggml-cpu.c.o] Error 1
../../../ggml/src/ggml-cpu/vec.cpp: In function 'void ggml_vec_dot_f32(int, float*, size_t, const float*, size_t, const float*, size_t, int)':
../../../ggml/src/ggml-cpu/vec.cpp:93:41: error: 'vsum' may be used uninitialized [-Werror=maybe-uninitialized]
   93 |         vsum = __riscv_vfmv_v_f_f32m8_tu(vsum, 0.0f, vl);
      |                ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
../../../ggml/src/ggml-cpu/vec.cpp:90:22: note: 'vsum' was declared here
   90 |         vfloat32m8_t vsum;
      |                      ^~~~
cc1plus: all warnings being treated as errors
make[2]: *** [ggml/src/CMakeFiles/ggml-cpu.dir/build.make:219: ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/vec.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:2199: ggml/src/CMakeFiles/ggml-cpu.dir/all] Error 2
```

`-DLLAMA_FATAL_WARNINGS=ON` has to be turned off for all the tests otherwise CI fails

(This can be created a separated issue. I can track the contributor and ping him if you want)

### Note 2

One RISC-V board may not be optimal for running all these tests so by the end of RISC-V summit North America, we are expecting more boards to arrive (around mid of November). So this PR can be treated as draft for reviews till then.

alitariq4589 committed:
Added RISC-V supported tests

alitariq4589 committed:
Added RISC-V supported tests

alitariq4589 committed:
Added default value for LLAMA_FATAL_WARNINGS and option to specify by user

alitariq4589 committed:
Added default value for LLAMA_FATAL_WARNINGS and option to specify by user

alitariq4589 committed:
Synced with origin

alitariq4589 committed:
Removed apt prompt

alitariq4589 commented:
@CISC There was a minor internet outage. It is fixed now. Can you check and let me know if the board still shows offline? If that is the case, I will restart the container.

> Any progress on this?

I had ordered and integrated 9 RISC-V boards with RVV1.0 so all tests can run smoothly, I am currently testing the ccache by running multiple tests if it works. Here is the [build](https://github.com/alitariq4589/llama.cpp/actions/runs/19536188023/job/56456001873). As soon as I am done testing, I will open the PR for review. Additionally I will also need a single runner token for adding all those boards in llama.cpp repository once I open the PR for review.

It would be great if we have some faster means of communication other than issues and emails. Do you use some other messaging platform (discord, mastodon etc.?). If it is okay, you can also join this [discord server](https://discord.gg/H7EGrzV93p).

CISC commented:
> @CISC There was a minor internet outage. It is fixed now. Can you check and let me know if the board still shows offline? If that is the case, I will restart the container.

I cancelled all the old jobs, but there are currently 2 new ones queued and not picked up yet.

> > Any progress on this?
> 
> I had ordered and integrated 9 RISC-V boards with RVV1.0 so all tests can run smoothly, I am currently testing the ccache by running multiple tests if it works. Here is the [build](https://github.com/alitariq4589/llama.cpp/actions/runs/19536188023/job/56456001873). As soon as I am done testing, I will open the PR for review. Additionally I will also need a single runner token for adding all those boards in llama.cpp repository once I open the PR for review.

Great, ping Georgi when you do.

> It would be great if we have some faster means of communication other than issues and emails. Do you use some other messaging platform (discord, mastodon etc.?). If it is okay, you can also join this [discord server](https://discord.gg/H7EGrzV93p).

Sorry, email only.

alitariq4589 committed:
Added RISC-V specific tests with corrections

Corrections included:
1. Changed the test names from debian to ubuntu as it is more stable than Debian Trixie
2. Added explicit compiler in cmake command as GCC compiler below version 14 have been recorded
to throw errors with rvv1.0 and some other extensions
3. Added dependencies which are not installed by default in the RISC-V Ubuntu 24.04
4. Separate ccache directory for all jobs as all the ccache results are not the same and may cause ccache to not work

alitariq4589 committed:
Resolved conflicts with master

alitariq4589 commented:
@CISC This PR is ready for review. I have excluded `ggml-ci-riscv64-native-cpu-low-perf` as it was causing boards to crash due to high resource utilization. There are 10x Banana Pi F3 boards arriving. When they do, I will use them for testing. 

The result of the added builds can be seen [here](https://github.com/alitariq4589/llama.cpp/actions/runs/19800721211) in my [fork](https://github.com/alitariq4589/llama.cpp). There are multiple attempts, so you can check each of them. As the number of boards integrated is greater than the number of tests ported for RISC-V (because there may be some builds in the queue to, these boards will offload some time), the ccache effect is not immediated visible in 4 attempts, but I have tested individually and so far, according to the stats, the ccache seems to be working (check [this](https://github.com/alitariq4589/llama.cpp/actions/runs/19763298033/job/56630083126) and [this](https://github.com/alitariq4589/llama.cpp/actions/runs/19785115477) job builds too for ccache results in which I executed each job multiple times for checking ccache results).

Following RISC-V boards will be integrated once I get the token.
- 8x Milk-V Jupiter boards
- 1x Banana Pi F3 (BPI F3) - This is already integrated with runner name `87a313462bbe`, and I will not add it again for duplication

For checking the resource utilization, I have set up grafana to track usage. This tracking site tracks the usage of the host machine and not the containers in which the builds will be running. Use the following links to view resource usage.
- [Summary of usage all the RISC-V RVV1.0 boards integrated with llama.cpp](https://monitor.cloud-v.co/d/b6c493c5-f573-4b02-ac62-5accb02db4af/all-risc-v-rvv1-0-machines-summary?orgId=2)
- [Individual tracking of all the RISC-V boards integrated with llama.cpp](https://monitor.cloud-v.co/dashboards/f/d7b6efcf-5448-42c3-988b-3a972e3527fd/llama-cpp)

_NOTE: Since our network engineer is out of the office for the next couple of days, `jupiter-16G-2`, `jupiter-16G-4`, `jupiter-16G-5` are not added in usage tracking, so you can ignore their shown data in the tracking dashboards (those are placeholders right now). I will add this within the next two weeks and will comment here._

@ggerganov, please share a github runner token when you can, and I will use that to register all these boards for builds. The token will be valid for one hour after generation, so let me know as soon as you generate it. You can send it to me at my email.

Let me know if anyone has any questions 🙂

ggerganov reviewed (commented):
@alitariq4589 Sending you the token in a min

ggerganov commented on ci/run.sh:
```diff
@@ -45,7 +45,11 @@ sd=`dirname $0`
+CMAKE_EXTRA="-DLLAMA_FATAL_WARNINGS=${LLAMA_FATAL_WARNINGS:-ON} -DLLAMA_CURL=ON"
+=======
 CMAKE_EXTRA="-DLLAMA_FATAL_WARNINGS=ON -DLLAMA_CURL=ON -DGGML_SCHED_NO_REALLOC=ON"
+>>>>>>> master
```
@xctan Could you take a look at this compile warning?

```
In file included from ../../../ggml/src/ggml-cpu/arch/riscv/quants.c:6:
../../../ggml/src/ggml-cpu/simd-mappings.h: In function 'riscv_compute_fp16_to_fp32':
../../../ggml/src/ggml-cpu/simd-mappings.h:101:9: error: ISO C does not support the '_Float16' type before C23 [-Werror=pedantic]
  101 |         _Float16 hf;
      |         ^~~~~~~~
../../../ggml/src/ggml-cpu/simd-mappings.h: In function 'riscv_compute_fp32_to_fp16':
../../../ggml/src/ggml-cpu/simd-mappings.h:108:9: error: ISO C does not support the '_Float16' type before C23 [-Werror=pedantic]
  108 |         _Float16 hf = (_Float16)f;
      |         ^~~~~~~~
../../../ggml/src/ggml-cpu/simd-mappings.h:108:24: error: ISO C does not support the '_Float16' type before C23 [-Werror=pedantic]
  108 |         _Float16 hf = (_Float16)f;
      |                        ^~~~~~~~
cc1: all warnings being treated as errors
```

alitariq4589 replied:
@ggerganov strangely I dont see conflicts after merging. I dont see conflicts here on the github ui too. Are you testing this on older commit? I just merged master branch to my branch and it seems okay.

CISC replied:
You have committed the merge conflict text, it just needs cleaning up.

alitariq4589 committed:
Merge branch 'master' into rvv1.0ci

alitariq4589 committed:
Resolved the merge conflict and cleaned up run.sh

alitariq4589 committed:
Update ci/run.sh

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

CISC commented:
> @CISC the `.github/workflows/build-riscv-native.yml` CI is failing. But should I remove this since most of the things are already covered in the added tests, and it seems redundant?

You tell me. :) If it is of no further value to you it can be removed.

alitariq4589 committed:
Removed previously added build ci for RISC-V

alitariq4589 committed:
Pulled branch 'rvv1.0ci' of github.com:alitariq4589/llama.cpp into rvv1.0ci

alitariq4589 committed:
Removed trailing whitespaces

alitariq4589 committed:
corrected build name

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

CISC committed:
cleanup

CISC commented on .github/workflows/build.yml:
```diff
@@ -1642,6 +1642,335 @@ jobs:
+            -DGGML_OPENMP=ON \
+            -DLLAMA_BUILD_EXAMPLES=ON \
+            -DLLAMA_BUILD_TOOLS=ON \
+            -DLLAMA_BUILD_TESTS=OFF \
```
```suggestion
            -DLLAMA_BUILD_TESTS=ON \
```

alitariq4589 committed:
Enabled build tests (1)

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

alitariq4589 committed:
Enabled build tests (2)

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

CISC commented on .github/workflows/build.yml:
```diff
@@ -1642,6 +1642,335 @@ jobs:
+        id: cmake_build
+        run: |
+          cmake -B build \
+            -DLLAMA_CURL=OFF \
```
```suggestion
            -DLLAMA_CURL=OFF \
            -DLLAMA_OPENSSL=ON \
```

CISC commented on .github/workflows/build.yml:
```diff
@@ -1642,6 +1642,335 @@ jobs:
+        id: cmake_build
+        run: |
+          cmake -B build \
+            -DLLAMA_CURL=OFF \
```
```suggestion
            -DLLAMA_CURL=OFF \
            -DLLAMA_OPENSSL=ON \
```

CISC committed:
enable openssl

CISC commented:
@alitariq4589 The runners seem to get randomly cancelled today, something up?

CISC commented:
@alitariq4589 Looks like the queue has gotten stuck (not auto-cancelled this time), runners offline?

alitariq4589 commented:
@CISC I have checked that some runners are still running and picking the jobs (e.g., jupiter-16G-9, jupiter-16G-8). I am pretty sure this is related to GitHub Actions. I will downgrade it to 2.328.0 from 2.329.0 to see if the issue persists.

alitariq4589 commented:
I have updated the github actions package versions inside the containers:

Mostly, which runner is picking up the job can also be observed from the usage dashboard as it uses all cores (dashboard link is in my above comment)

Following are the runner versions (some boards were working fine - such as jupiter-16G-9, they were picking jobs as usual so i didnt downgrade them)

- BPF3-16G-1: version 2.321.0 - I have some cancellations here but didnt change its version yet as cancellations are rare in the logs
- jupiter-16G-1: version 2.328.0 - Version has been downgraded as it was not picking up jobs in the logs, now it is picking up the jobs
- jupiter-16G-2: version 2.328.0 - Version has been downgraded as it was not picking up jobs in the logs, now it is picking up the jobs
- jupiter-16G-4: version 2.329.0 - Version not changed as it is not being monitored right now, will observe it once it is also being tracked
-  jupiter-16G-5: version 2.329.0 - Version not changed as it is currently offline
- jupiter-16G-6: version 2.329.0: Version not changed as it is picking up jobs just fine
- jupiter-16G-7: version 2.328.0 - Version has been downgraded as it was not picking up jobs in the logs, now it is picking up the jobs
- jupiter-16G-8: version 2.328.0 - Version has been downgraded as it was not picking up jobs in the logs, now it is picking up the jobs
- jupiter-16G-9: version 2.329.0: Version not changed as it is picking up jobs just fine


Let's see how it works now.

CISC commented:
@alitariq4589 Are all of them active right now? It looks there are only 5 picking up jobs.

alitariq4589 commented:
> @alitariq4589 Are all of them active right now? It looks there are only 5 picking up jobs.

@CISC Sorry for the delayed response. I was outside at the time and couldn't reply. But at that time I checked all the boards were picking up the jobs and in the usage, I also see 75.3% of CPU usage of all boards collectively. So everything was working okay. 

One thing to note is that, I setup a refresh script which checks if the boards have lost connection from github and refresh it. This script polls every 3 minutes to see if there is any anomaly. So there can be at most 3 minutes delay but not more than that before a board starts picking the jobs again.

CISC commented:
@alitariq4589 Whenever you have time please run `git lfs install` on `jupiter-16G-4` (I assume this is the one that is failing tokenizers test atm). :)

CISC commented:
> @CISC Has there been any unwanted job cancellations or anomalies in the RISC-V CI?

They do happen somewhat frequently, the most recent one being this I think:
https://github.com/ggml-org/llama.cpp/actions/runs/21952241958/job/63405997848

alitariq4589 commented:
@CISC I am sorry for the inconvenience. I will now upgrade the runner to the latest version, and (hopefully) all these issues will be resolved. 

According to what I assessed in the past couple of days, these are happening because of the following discrepancies.

1. The Ubuntu image used in the dockerfile which is running as the runner is [riscv64/ubuntu](https://hub.docker.com/u/riscv64), and that is not the official image. There IS one official image from Ubuntu now. But updating the image will cause a little downtime
2. The Github runner version that is running is now older than the upstream version, so it tries to install the latest version of github runner. But, since that is not available forthe  RISC-V architecture, it interrupts the running job, cancelling it in the process. That is why the job cancellations are now more frequent. I have not yet implemented a patch for this. The best way is to keep the github runner up to date. 
3. The older versions of .NET silently crash internally with segfault errors because of bugs. The newer version (10.0.102) will hopefully work fine.

I will start the upgrade process. Hopefully, there won't be any interruptions as I will upgrade the runners one by one. I should be done in 5-6 hours if no problem is faced.

I will add a comment once I am done with this.

CISC commented:
> I would also like to mention that pytorch is now released for riscv. I have set up a workflow which fetches the source code from official upstream, builds and releases it without adding any kind of change. That means the RISC-V workflows which require pytorch for RISC-V can now be added.

Nice, thanks for following up. :)

alitariq4589 commented:
@CISC, I noticed that there are several jobs in the queue. I was waiting for the runner to complete all the jobs so I can patch it, but looking at the logs, i dont think that is going to happen :sweat_smile: . So if I try to patch the runner in this state, the running job will be cancelled. 

Is it okay if I cancel the job running on the boards to patch them? You can also tell me a specific time to upgrade the packages when you think it will not affect the CI and PRs considerably.

CISC commented:
> @CISC, I noticed that there are several jobs in the queue. I was waiting for the runner to complete all the jobs so I can patch it, but looking at the logs, i dont think that is going to happen 😅 . So if I try to patch the runner in this state, the running job will be cancelled.

Yeah, quite unlikely at this point. :D

> Is it okay if I cancel the job running on the boards to patch them? You can also tell me a specific time to upgrade the packages when you think it will not affect the CI and PRs considerably.

Just go ahead, if it's crucial for a PR we can rerun them.

alitariq4589 commented:
I also noticed that two runners were not properly upgraded. I have updated their file,s and they will automatically restart once no jobs are running.
